File: 01-beginner/02-ros2-core-concepts/demo/cpp/src/listener.cpp
// Subscribes to /chatter and logs every message it receives.
//
// C++ equivalent of demo/python/listener.py.

#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using std::placeholders::_1;  // used below to bind the callback's one argument

class Listener : public rclcpp::Node
{
public:
  Listener() : Node("listener")
  {
    // create_subscription<MsgType>(topic_name, qos_depth, callback).
    // std::bind(&Listener::on_message, this, _1) creates a callable that,
    // when invoked with one argument, calls this->on_message(that_arg) —
    // the C++ way of registering a member function as a callback.
    subscription_ = this->create_subscription<std_msgs::msg::String>(
      "chatter", 10, std::bind(&Listener::on_message, this, _1));
  }

private:
  void on_message(const std_msgs::msg::String & msg) const
  {
    RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg.data.c_str());
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Listener>());
  rclcpp::shutdown();
  return 0;
}

File: 01-beginner/02-ros2-core-concepts/demo/python/listener.py
#!/usr/bin/env python3
"""Subscribes to /chatter and logs every message it receives."""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class Listener(Node):
    def __init__(self):
        super().__init__("listener")

        # create_subscription(msg_type, topic_name, callback, queue_size).
        # The callback (_on_message) is invoked by the executor whenever
        # a message arrives — we never call it ourselves, and we never
        # poll for new data; ROS2's spin() loop delivers it to us.
        #
        # Note there's no code here that knows a "talker" node exists.
        # This subscriber would work identically if a completely
        # different program published on /chatter with the right message
        # type — that decoupling is the whole point of pub/sub.
        self._subscription = self.create_subscription(
            String, "chatter", self._on_message, 10
        )

    def _on_message(self, msg: String):
        self.get_logger().info(f'I heard: "{msg.data}"')


def main():
    rclpy.init()
    node = Listener()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
